[Layout] Harden RenderGrid::gridAreaRangeForOutOfFlow() code for scenarios we haven't completed layout

zakariaridouh opened the pull request:
#### c4a30a15887b1eef4150b93576e7527d0b574e6a
<pre>
[Layout] Harden RenderGrid::gridAreaRangeForOutOfFlow() code for scenarios we haven&apos;t completed layout
<a href="https://bugs.webkit.org/show_bug.cgi?id=300934">https://bugs.webkit.org/show_bug.cgi?id=300934</a>
<a href="https://rdar.apple.com/162810289">rdar://162810289</a>

Reviewed by Sammy Gill.

In some situations, we can get into a state where we are attempting to access the `positions` vector in
but haven&apos;t completed in-flow placement and grid sizing yet.

We should add an assert to verify we&apos;re not doing anything improperly previously in the code, and return something basic
that doesnt access unbuilt data structures on production builds.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::gridAreaRangeForOutOfFlow const):

Canonical link: <a href="https://commits.webkit.org/302263@main">https://commits.webkit.org/302263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c329a6f1c17629c3dd10bfa2a0432a24d0ce71cf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/126702 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/46344 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/37308 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/133661 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/78346 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9182e74a-c32d-41eb-b32d-adee63df2826) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/128573 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/46973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/54879 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/96415 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65065 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ebfd4739-d71b-4716-a0c9-9127d9ef9009) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/129650 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/37583 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/113316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/76940 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a05c5a95-45ec-4ad1-ac66-26cac4b010bd) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/36469 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/31500 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/77058 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/107399 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/31800 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/136236 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/53383 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41073 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/104929 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/53874 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/109677 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/104630 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27054 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50123 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/28457 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/50823 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/53308 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/59111 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/52574 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/55910 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/54316 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -2250,7 +2250,13 @@ LayoutRange RenderGrid::gridAreaRangeForOutOfFlow(const RenderBox& gridItem, Sty
     auto& positions = this->positions(direction);
+    if (positions.isEmpty()) {
+        ASSERT_WITH_SECURITY_IMPLICATION(!positions.isEmpty());
+        return LayoutRange(borderEdge, isRowAxis ? clientLogicalWidth() : clientLogicalHeight());
```
I think this could probably be return `{ borderEdge, isRowAxis ? clientLogicalWidth() : clientLogicalHeight() };`

zakariaridouh replied:
Other parts of the code in this function use the `return LayoutRange(borderEdge, isRowAxis ? clientLogicalWidth() : clientLogicalHeight());` style, why use the `{}` style?

sammygill replied:
For me I think it's nicer because there's less noise/typing around the crux of the code. There are probably also some technical differences but I don't think they really apply here. FWIW you can use either I think it's fine either way and maybe it's better to just match the pattern of the existing code in scenarios like this where there is no functional difference

zakariaridouh replied:
Ah that's fair, thanks for the explanation Sammy =)

zakariaridouh committed:
[Layout] Harden RenderGrid::gridAreaRangeForOutOfFlow() code for scenarios we haven't completed layout
https://bugs.webkit.org/show_bug.cgi?id=300934
rdar://162810289

Reviewed by Sammy Gill.

In some situations, we can get into a state where we are attempting to access the `positions` vector in
but haven't completed in-flow placement and grid sizing yet.

We should add an assert to verify we're not doing anything improperly previously in the code, and return something basic
that doesnt access unbuilt data structures on production builds.

* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::gridAreaRangeForOutOfFlow const):

Canonical link: https://commits.webkit.org/302263@main
